documentation/new-version-guideline.md

```diff
@@ -0,0 +1,302 @@
+
+# Azure Go Management SDK Guideline
+
+Azure Go management SDK follows the [new Azure SDK guidelines](https://azure.github.io/azure-sdk/general_introduction.html), try to create easy-to-use APIs that are idiomatic, compatible, and dependable.
+
+You can find the full list of management modules [here](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk).
+
+In this guideline, we will give some instructions about the API usage pattern as well as trouble shooting method. For those are new to management Go SDK, please refer to [quickstart](./new-version-quickstart.md). For those migrate from older versions of management Go SDK, please refer to [migration guide](https://aka.ms/azsdk/go/mgmt/migration).
+
+## Pageable Operations
+
+### General usage
+
+Pageable operations return final data over multiple GET requests. Each GET will receive a page of data consisting of a slice of items. You need to use New*Pager to create a pager helper for all pageable operations. With the returned *runtime.Pager[T], you could fetch pages and determine if there are more pages to fetch. For examples:
+
+```go
+pager := rgClient.NewListPager(nil)
+var resourceGroups []*armresources.ResourceGroup
+for pager.More() {
+    nextResult, err := pager.NextPage(ctx)
+    if err != nil {
+        // handle error...
+    }
+    if nextResult.ResourceGroupListResult.Value != nil {
+        resourceGroups = append(resourceGroups, nextResult.ResourceGroupListResult.Value...)
+    }
+}
+// dealing with `resourceGroups`
+```
+
+> NOTE: No IO calls are made until the NextPage() method is invoked. The read consistency across pages is determined by the service implement.
+
+### Item iterator
+
+If you are not care about the underlaying detail about the pageable operation, you could use the following generic utility to create a per-item iterator for all pageable operation.
+
+***Item iterator utility***
+```go
+type PageConstraint[TItem any] interface {
+	Items() []*TItem
+}
+
+type Iterator[TItem any, TPage PageConstraint[TItem]] struct {
+	pager *runtime.Pager[TPage]
+	cur   []*TItem
+	index int
+}
+
+func (iter *Iterator[TItem, TPage]) More() bool {
+	return iter.pager.More() || iter.index < len(iter.cur)
+}
+
+func (iter *Iterator[TItem, TPage]) NextItem(ctx context.Context) (*TItem, error) {
+	if iter.index == len(iter.cur) && !iter.pager.More() {
+		return nil, errors.New("no more items")
+	}
+	if iter.cur == nil || iter.index == len(iter.cur) {
+		// first page or page exhausted
+		page, err := iter.pager.NextPage(ctx)
+		if err != nil {
+			return nil, err
+		}
+		iter.cur = page.Items()
+		iter.index = 0
+	}
+	item := iter.cur[iter.index]
+	// advance item
+	iter.index++
+	return item, nil
+}
+
+func NewIterator[TItem any, TPage PageConstraint[TItem]](pager *runtime.Pager[TPage]) *Iterator[TItem, TPage] {
+	return &Iterator[TItem, TPage]{
+		pager: pager,
+	}
+}
+```
+
+***Usage***
+```go
+iter := NewIterator[*armresources.ResourceGroup](rgClient.NewListPager(nil))
```

Comment:
Can the compiler infer the generic parameter `*armresources.ResourceGroup`? This looks a bit too verbose

Reply:
Unfortunately not because of `New*Pager` also return an generic type.